### PR TITLE
Upgrade to Java 17

### PR DESCRIPTION
**Pull request by RSHKwee**
An upgrade to Java 17 (OpenJDK 17) is performed.
OFX Settings is introduced for storage of Encoding and Currency.

### CODE REVIEW

File: src/main/java/com/webcohesion/ofx4j/client/impl/FIDataList.java [hunk 1]
```diff
@@ -16,8 +16,9 @@
 
 package com.webcohesion.ofx4j.client.impl;
 
-import javax.xml.bind.annotation.XmlElement;
-import javax.xml.bind.annotation.XmlRootElement;
+//import javax.xml.bind.annotation.XmlElement;
+//import javax.xml.bind.annotation.XmlRootElement;
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/client/impl/FIDataList.java]: Please remove commented code.

File: src/main/java/com/webcohesion/ofx4j/client/impl/LocalResourceFIDataStore.java [hunk 2]
```diff
@@ -21,8 +21,9 @@
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 
-import javax.xml.bind.JAXBContext;
-import javax.xml.bind.JAXBException;
+//import javax.xml.bind.JAXBContext;
+//import javax.xml.bind.JAXBException;
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/client/impl/LocalResourceFIDataStore.java]: Please remove commented code.

File: src/main/java/com/webcohesion/ofx4j/domain/data/common/Currency.java [hunk 3]
```diff
@@ -18,26 +18,32 @@
...
  * @author Ryan Heaton
  * @see "Section 5.2, OFX Spec"
  */
-@Aggregate ( "CURRENCY" )
+@Aggregate("CURRENCY")
 public class Currency {
-
   private String code = java.util.Currency.getInstance(Locale.US).getCurrencyCode().toUpperCase();
   private Float exchangeRate;
+  private OFXSettings m_ofxSettings;
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/domain/data/common/Currency.java]: Please remove the `m_` from field names to align better with project codestyle.

File: src/main/java/com/webcohesion/ofx4j/domain/data/common/StatementResponse.java [hunk 4]
```diff
@@ -33,14 +34,21 @@ public abstract class StatementResponse extends ResponseMessage implements Accou
   private BalanceInfo ledgerBalance;
   private BalanceInfo availableBalance;
   private String marketingInfo;
+  private OFXSettings m_ofxSettings;
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/domain/data/common/StatementResponse.java]: Please remove the `m_` from field names to align better with project codestyle.

File: src/main/java/com/webcohesion/ofx4j/domain/data/profile/CoreMessageSetInfo.java [hunk 5]
```diff
@@ -41,13 +42,19 @@ public class CoreMessageSetInfo {
   private SynchronizationCapability syncCapability;
   private Boolean fileBasedErrorRecoverySupport;
   private Integer timeout;
+  private OFXSettings m_ofxSettings;
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/domain/data/profile/CoreMessageSetInfo.java]: Please remove the `m_` from field names to align better with project codestyle.

File: src/main/java/com/webcohesion/ofx4j/domain/data/profile/CoreMessageSetInfo.java [hunk 6]
```diff
@@ -62,19 +69,23 @@ public void setVersion(String version) {
   }
 
   /**
-   * The name of the service provider (sometimes the message set processing is outsourced).
+   * The name of the service provider (sometimes the message set processing is
+   * outsourced).
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/domain/data/profile/CoreMessageSetInfo.java]: Please remove the unnecessary line break changes. (Here and below.)

File: src/main/java/com/webcohesion/ofx4j/domain/data/signon/SignonResponse.java [hunk 7]
```diff
@@ -45,11 +46,19 @@ public class SignonResponse extends ResponseMessage implements StatusHolder {
   private FinancialInstitution financialInstitution;
   private String sessionId;
   private String accessKey;
+  private OFXSettings m_ofxSettings;
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/domain/data/signon/SignonResponse.java]: Please remove the `m_` from field names to align better with project codestyle.

File: src/main/java/com/webcohesion/ofx4j/domain/data/signon/SignonResponse.java [hunk 8]
```diff
@@ -174,15 +184,16 @@ public void setProfileLastUpdated(Date profileLastUpdated) {
...
   public Date getAccountLastUpdated() {
     return accountLastUpdated;
   }
 
   /**
    * The date/time that the user's account information was updated.
    *
-   * @param accountLastUpdated The date/time that the user's account information was updated.
+   * @param accountLastUpdated The date/time that the user's account information
+   *                           was updated.
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/domain/data/signon/SignonResponse.java]: Please remove the unnecessary line break changes. (Here and below.)

File: src/main/java/com/webcohesion/ofx4j/io/v1/OFXV1Writer.java [hunk 9]
```diff
@@ -27,17 +28,28 @@
  * @author Ryan Heaton
  */
 public class OFXV1Writer implements OFXWriter {
+  private OFXSettings m_ofxSettings = OFXSettings.getInstance();
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/io/v1/OFXV1Writer.java]: Please remove the `m_` from field names to align better with project codestyle.

File: src/main/java/com/webcohesion/ofx4j/io/v1/OFXV1Writer.java [hunk 10]
```diff
@@ -27,17 +28,28 @@
...
 
   private static final String LINE_SEPARATOR = "\r\n";
   protected boolean headersWritten = false;
   protected final Writer writer;
-  private boolean writeAttributesOnNewLine = false;
+  private boolean writeAttributesOnNewLine = m_ofxSettings.getWriteAttributesOnNewLine();
+  private String m_Filename = "";
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/io/v1/OFXV1Writer.java]: Please remove the `m_` from field names to align better with project codestyle.

File: src/main/java/com/webcohesion/ofx4j/io/v2/OFXV2Writer.java [hunk 11]
```diff
@@ -27,27 +28,42 @@
  * @author Ryan Heaton
  */
 public class OFXV2Writer extends OFXV1Writer {
+  private OFXSettings m_ofxSettings;
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/io/v2/OFXV2Writer.java]: Please remove the `m_` from field names to align better with project codestyle.

File: src/main/java/com/webcohesion/ofx4j/io/BaseOFXReader.java [hunk 12]
```diff
@@ -39,6 +41,7 @@ public abstract class BaseOFXReader implements OFXReader {
   private static final Log LOG = LogFactory.getLog(BaseOFXReader.class);
   public static final Pattern OFX_2_PROCESSING_INSTRUCTION_PATTERN = Pattern.compile("<\\?OFX ([^\\?]+)\\?>");
   private OFXHandler contentHandler = new DefaultHandler();
+  private OFXSettings m_ofxSettings = OFXSettings.getInstance();
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/io/BaseOFXReader.java]: Please remove the `m_` from field names to align better with project codestyle.

File: src/main/java/com/webcohesion/ofx4j/io/BaseOFXReader.java [hunk 13]
```diff
@@ -84,7 +90,8 @@ public void parse(Reader reader) throws IOException, OFXParseException {
     int ch = reader.read(buffer);
     while ((ch != -1) && (!Arrays.equals(buffer, firstElementStart))) {
       if (!contains(buffer, '<')) {
-        //if the buffer contains a '<', then we might already have marked the beginning.
+        // if the buffer contains a '<', then we might already have marked the
+        // beginning.
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/io/BaseOFXReader.java]: Please remove the unnecessary line break changes. (Here and below.)

File: src/main/java/com/webcohesion/ofx4j/OFXSettings.java [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+import com.webcohesion.ofx4j.generated.CurrencyEnum;
+
+/**
+ * OFXSettings some general parameter definitions. This is a singleton object.
+ * 
+ * @author RSH Kwee
+ *
+ */
+public class OFXSettings {
+  private static OFXSettings uniqueInstance;
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/OFXSettings.java:15]: Please use ALL_CAPS_SNAKE_CASE to better align with project codestyle.

File: src/main/java/com/webcohesion/ofx4j/OFXSettings.java [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+
+  private String m_Encoding = "ISO-8859-1";
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/OFXSettings.java]: Please remove the `m_` from field names to align better with project codestyle. (Here and below.)

File: src/main/java/com/webcohesion/ofx4j/OFXSettings.java [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+  private String m_Currency = "EUR";
+
+  /*  
+  @formatter:off
+ 
+  @formatter:on
```
Comment (stoicflame) [inline src/main/java/com/webcohesion/ofx4j/OFXSettings.java]: Please remove the `formatter` comments. (What are those, anyway?)

File: src/test/java/com/webcohesion/ofx4j/io/v1/TestOFXV1Writer.java [hunk 17]
```diff
@@ -29,12 +31,14 @@ public class TestOFXV1Writer extends TestCase {
    * tests character escaping
    */
   public void testCharacterEscaping() throws Exception {
+    OFXSettings OFXSetting = OFXSettings.getInstance();
```
Comment (stoicflame) [inline src/test/java/com/webcohesion/ofx4j/io/v1/TestOFXV1Writer.java:34]: Please use lowercase variable naming to better align with project codestyle (e.g. `ofxSetting`).

File: pom.xml [hunk 18]
```diff
@@ -6,26 +6,27 @@
...
 
-  <version>1.36</version>
+  <version>1.37</version>
   <name>ofx4j</name>
   <description>
     OFX4J is a Java implementation of Open Financial Exchange, which defines web service
     APIs for interfacing with financial institutions.
   </description>
-  <url>https://github.com/stoicflame/ofx4j</url>
+<!--  <url>https://github.com/stoicflame/ofx4j</url> -->
```
Comment (stoicflame) [inline pom.xml:15]: Please restore GH url.

File: pom.xml [hunk 19]
```diff
@@ -6,26 +6,27 @@
...
+  <url>https://github.com/rshkwee/ofx4j</url>
 
   <developers>
     <developer>
       <name>Ryan Heaton</name>
     </developer>
   </developers>
-
+  <!--
```
Comment (stoicflame) [inline pom.xml:23]: Please restore GH url.

File: pom.xml [hunk 20]
```diff
@@ -35,16 +36,42 @@
 
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <maven.test.failure.ignore>true</maven.test.failure.ignore>
```
Comment (stoicflame) [inline pom.xml]: Please remove unused property.

File: pom.xml [hunk 21]
```diff
@@ -35,16 +36,42 @@
...
   </properties>
 
   <build>
     <plugins>
+      <plugin>
+	    <groupId>org.codehaus.mojo</groupId>
+		<artifactId>jaxb2-maven-plugin</artifactId>
```
Comment (stoicflame) [inline pom.xml:45]: Please align tab spacing for better readability.

File: pom.xml [hunk 22]
```diff
@@ -35,16 +36,42 @@
...
+	    </configuration>
+	  </plugin>
+    
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-compiler-plugin</artifactId>
         <configuration>
-          <source>1.6</source>
-          <target>1.6</target>
+          <source>17</source>
```
Comment (stoicflame) [inline pom.xml]: I'd like to use the lowest common denominator to reduce upgrade friction. Can we get this to build in Java 8? If not at least Java 11?

File: pom.xml [hunk 23]
```diff
@@ -176,7 +203,16 @@
       <!--optional (only for the command-line utilities-->
       <optional>true</optional>
     </dependency>
-
+    <dependency>
+      <groupId>com.sun.xml.bind</groupId>
+      <artifactId>jaxb-impl</artifactId>
```
Comment (stoicflame) [inline pom.xml]: What still needs the `jaxb-impl` dependency?

File: pom.xml [hunk 24]
```diff
@@ -187,31 +223,31 @@
   </dependencies>
 
   <distributionManagement>
+ <!--  
     <snapshotRepository>
```
Comment (stoicflame) [inline pom.xml]: Please restore the ossrh repository.

**Review by stoicflame — CHANGES_REQUESTED** [review]
Thanks for the PR! Please address the comments I left.

Also please remove `Locale.xlsx` from the PR.

File: src/main/java/com/webcohesion/ofx4j/OFXSettings.java (hunk 16 above)
Comment (RSHKwee) [reply src/main/java/com/webcohesion/ofx4j/OFXSettings.java]: Its a marker for Eclips to instruct Eclips not to format (automaticly) the code between the markers.

File: pom.xml (hunk 23 above)
Comment (RSHKwee) [reply pom.xml]: Not needed, indeed. Have removed it.

File: pom.xml (hunk 22 above)
Comment (RSHKwee) [reply pom.xml]: Have changed it in "1.8".

**Comment by RSHKwee** [comment]
I have processed your comments, I hope that I have processed all your comments.....

**Comment by stoicflame** [comment]
Thanks for the PR! This has been merged and included in [version 1.37](https://github.com/stoicflame/ofx4j/releases/tag/v1.37).

**Pull request merged**